flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -375,4 +397,29 @@ void onConsumedSubpartition(int subpartitionIndex) {
 	private void checkInProduceState() throws IllegalStateException {
 		checkState(!isFinished, "Partition already finished.");
 	}
+
+	/**
+	 * Check whether all subpartitions' backlogs are less than the limitation of max backlogs, and make this partition
+	 * available again if yes.
+	 */
+	public void notifyDecreaseBacklog(int buffersInBacklog) {
+		if (buffersInBacklog == maxBuffersPerChannel) {
+			if (--unavailableSubpartitionsCount == 0) {
+				CompletableFuture<?> toNotify = availabilityHelper.getUnavailableToResetAvailable();
+				toNotify.complete(null);
+			}
+		}
+	}
+
+	/**
+	 * Check whether any subpartition's backlog exceeds the limitation of max backlogs, and make this partition
+	 * unavailabe if yes.
+	 */
+	public void notifyIncreaseBacklog(int buffersInBacklog) {
+		if (buffersInBacklog == maxBuffersPerChannel + 1) {
+			if (++unavailableSubpartitionsCount == 1) {
+				availabilityHelper.resetUnavailable();
+			}
+		}
+	}
```

Comment:
Continuing [the previous thread ](**url**)
> each PipelinedSubpartition has it's own buffers field, so the updates here to unavailableSubpartitionsCnt are complete not synchronized. Some simple solution might be to change unavailableSubpartitionsCnt to an AtomicLong, but let me think about alternative solutions.

There is also another issue. `AvailabilityHelper` is not thread safe class, while you are accessing and modifying it's state from different threads, so my previous idea with the `AtomicLong unavailableSubpartitionsCount ` wouldn't work. This also affects `getAvailableFuture`. 

I don't know if that can be solved in an efficient way. One thing that could help a bit, is some old idea of mine
https://github.com/pnowojski/flink/commit/24e360e7db2d6f354c2f7369b95607eee1a5cb2c
that there is no reason for each subpartition to have unique lock. Having one global per partition lock would help with couple of issues, like `flushAll()`, some load rebalancing ideas that we had or here. However we would have to also synchronize `ResultPartition#getAvailableFuture(...)` which would affect performance on the happy/hot path :/

Maybe a solution would be to not use `availabilityHelper` here at all. But instead `availabilityHelper.resetUnavailable();` enqueue an action in the mailbox, which would suspend default action and instead of `toNotify.complete(null);`, also enqueue a mailbox action to re-enable default actions. That would increase a cost of `#notifyIncreaseBacklog` call, but not on the happy/hot path, but only if this task is being back-pressured, so when performance shouldn't matter.

However we would still need either `AtomicLong` or global `ResultPartition.lock` to not have race conditions on `unavailableSubpartitionsCount`.

(lets think this though before jumping to implementation)

Reply:
Ok, I've just thought about probably a simpler and better idea. Why don't we move this anti data skew logic inside the `LocalBufferPool`? It's already synchronized for recycling the memory segments from netty threads and requesting from the task thread and it already implements `AvailabilityProvider` with an `AvailabilityHelper`. It currently doesn't know which channel/subpartition assigned to which buffer but it might be doable to pass this knowledge there:

- add subpartition index argument to `ResultPartitionWriter#getBufferBuilder`
- embed it into the `NetworkBuffer` inside `LocalBufferPool#toBuffer`? Maybe via passing a custom `BufferRecycler`?
- use the subpartition id during recycling

Couple of caveats:

- `LocalBufferPool` is also used for input gates at least, so it might need to support managing buffers with and without channels accounting.
- for `BroadcastRecordWriter` the accounting would either need to be disabled, or `Integer.MAX_VALUE` could be used to indicate to account all channels?
- there might be some other code paths leading to `LocalBufferPool` that do not care about anti data skew accounting.

Reply:
Mmm.. it seems to make sense, but let me make sure I understand this correctly.
Firstly we maintain a `int[] subpartitionsBufferCount` in `LocalBufferPool` to represent the count of each subpartition's buffers.
(1) About adding buffers. 
	a. We add an new version of `requestMemorySegment()`, maybe something like `requestMemorySegment(int subpartitionIndex)` and operate the  `subpartitionsBufferCount` inside the new function, which will only be called by `ResultPartition` without affecting the calling stack of `InputGate`.
	b. Add a channel index attribute into `NetworkBuffer` to make sure we know where the buffer comes from when recycling.
(2) About recycling. Create a custom `BufferRecyler` with channel information like `new CustomBufferRecycler(subpartitionIndex)` to make sure we know where the buffer comes from when recycling. And this means we need to create a single `BufferRecycler` instance for every subpartition. If I understand this part correctly, wouldn't it be adding too many (thousands of) `BufferRecycler` instances here?

Reply:
Yes, more or less that was my last idea (keep in mind that I might be missing something).

>  If I understand this part correctly, wouldn't it be adding too many (thousands of) BufferRecycler instances here?

For one thing, we could create one instance of `CustomBufferRecycler` per subpartitionIndex and keep it referenced on some array field `CustomBufferRecycler[] recyclers` - we don't have to keep creating them on each request - so it shouldn't affect performance/GC pressure.

Secondly, we already have per subpartition index multiple classes (for example `PipelinedSubpartition`, `PipelinedSubpartitionView`, `RemoteInputChannel`, `BufferBuilder`, `BufferConsumer` and all of theirs fields + buffers/segments themselves). One more shouldn't matter?

Thirdly, CustomBufferRecycler should be ~16 bytes, that's just 160KB for 10000 channels.

Lastly, we could also hide `subpartitionIndex` inside `NetworkBuffer` class, but I think that might be less clean taking into account that `InputGate` won't be using this field? But I might be wrong/persuaded the other direction here.

Reply:
@Jiayi-Liao just a small heads up. There is another PR that's in progress which is touching similar places (take a look at `RecordWriter#getBufferBuilder` in here https://github.com/apache/flink/pull/11564) - it's adding non blocking variant of `requestingBufferBuilder` from the `LocalBufferPool` up to the `RecordWriter`. It will be probably merged before your PR, so you will probably have some some conflicts during one of the rebase, that should be not that difficult to resolve, but just keep it in mind.

Reply:
@pnowojski Thanks for reminding :). A new idea occurs to me this morning, maybe it'll help open our thoughts a little bit. Since each subpartition is synchronized on buffers, what if we let each subpartition know its buffers' limit and we create the availability future in every subpartition? A subpartition will be unavailable when number of buffers are increasing and exceeds the limit, will be available when number of buffers are decreasing. In `ResultPartition`, there'll be something like this:

```
public boolean addBufferConsumer(BufferConsumer bufferConsumer, int subpartitionIndex) {
        ...
        ...
        subpartition.add(bufferConsumer);

        if (!subpartition.isAvailabe()) {
                subpartitionsFuture.combine(subpartition.availableFuture);
        }
        ...
}
```

But the problem here is we may have hundreds of futures combined together in the subpartitionsFuture if user sets a low value to the number of max buffers.

Reply:
Yes I guess you are right.

> But the problem here is we may have hundreds of futures combined together in the subpartitionsFuture if user sets a low value to the number of max buffers.

I think that wouldn't be a serious issue. Once a single sub-partition is backpressured because of data skew, data processing should pause. So as long as we are respecting the availability, there should be at most one blocked supbartition. 

But it would be an issue nonetheless. It could happen for example for flatMap operator, when it's producing a lot of output on a single call.

So the `subpartitionsFuture` would have to maintain a `List<CompletableFuture>` to keep track of all of the blocked subpartitions. With this in mind, I'm not entirely sure if this approach would be easier/better than going through `LocalBufferPool`? Maybe it would be better to have a single source of truth for output availability (`LocalBufferPool`), instead of distributing the logic among `LocalBufferPool` and subpartitions?

Reply:
Okay. I'm going to implement this in your way. This will introduce a few function changes in LocalBufferPool:

1. Interface changes in BufferProvider. Add `subpartitionIndex` as a parameter to `requestBufferBuilder` and `requestBufferBuilderBlocking` so that `BufferPool` knows which subpartition the request comes from.
2. Add `requestMemorySegment(int subpartitionIndex)` and `recycle(int subpartitionIndex)`, and new function will reuse most logic of the original function, and the original one will be like this:

```
private MemorySegment requestMemorySegment() throws IOException {
	return requestMemorySegment(INVALID_SUBPARTITION_INDEX);
}
```

3. Introduce class `SubpartitionBufferRecycler`. 

```
class SubpartitionBufferRecycler implements BufferRecycler {

	private int subpartitionIndex;
	private LocalBufferPool bufferPool;

	SubpartitionBufferRecycler(int subpartitionIndex, LocalBufferPool bufferPool) {
		this.subpartitionIndex = subpartitionIndex;
		this.bufferPool = bufferPool;
	}

	@Override
	public void recycle(MemorySegment memorySegment) {
		bufferPool.recycle(memorySegment, subpartitionIndex);
	}
}
```

Reply:
1. 👍 
2. `INVALID_SUBPARTITION_INDEX` -> `UNKNOWN_SUBPARTITION_INDEX`?

I guess it would be nicer to avoid using some magic constants, but to just keep `requestMemorySegment()` without the `subpartitionIndex` parameter besides a version with the parameter, but I don't see how this could be done without duplicating quite a bit of code.
3. 👍

Reply:
@pnowojski  `InputChannel` and `Subpartition` share the function `requestMemorySegment()`, so it looks like we don't have another way except using `UNKKNOW_SUBPARTITION_INDEX` if we want to reuse the logic in original function. BTW,  I've updated the description and the codes, perhaps you can take a look.